File: tests/test_parser.py
import sys
import os
import glob
import StringIO
import unittest
import new

def parseTestcase(testString):
    testString = testString.split("\n")
    try:
        if testString[0] != "#data":
            print testString
        assert testString[0] == "#data"
    except:
        raise
    input = []
    output = []
    errors = []
    currentList = input
    for line in testString:
        if line and not (line.startswith("#errors") or
          line.startswith("#document") or line.startswith("#data")):
            if currentList is output:
                if line.startswith("|"):
                    currentList.append(line[2:])
                else:
                    currentList.append(line)
            else:
                currentList.append(line)
        elif line == "#errors":
            currentList = errors
        elif line == "#document":
            currentList = output
    return "\n".join(input), "\n".join(output), errors

def convertTreeDump(treedump):
    """convert the output of str(document) to the format used in the testcases"""
    treedump = treedump.split("\n")[1:]
    rv = []
    for line in treedump:
        if line.startswith("|"):
            rv.append(line[3:])
        else:
            rv.append(line)
    return "\n".join(rv)

class TestCase(unittest.TestCase):
    def runParserTest(self, input, output, errors):
        import parser
        #XXX - move this out into the setup function
        #concatenate all consecutive character tokens into a single token
        p = parser.HTMLParser()
        document = p.parse(StringIO.StringIO(input))
        errorMsg = "\n".join(["\n\nExpected:", output, "\nRecieved:",
          convertTreeDump(document.printTree())])
        self.assertEquals(output, convertTreeDump(document.printTree()),
          errorMsg)

def test_parser():
    for filename in glob.glob('tree-construction/*.dat'):
        f = open(filename)
        test = []
        documentSeen = False
        for line in f:
            # XXX This algorithm would need to be changed if we want to get rid
            # of the double newline requirement at the end of test files.
            if line.startswith("#document"):
                documentSeen = True
            if not line == "\n":
                test.append(line[:-1])
            elif line == "\n" and not documentSeen:
                test.append(line[:-1])
            else:
                input, output, errors = parseTestcase("\n".join(test))
                yield TestCase.runParserTest, input, output, errors
                test = []
                documentSeen = False

def buildTestSuite():
    tests = 0
    for func, input, output, errors in test_parser():
        tests += 1
        testName = 'test%d' % tests
        testFunc = lambda self, method=func, input=input, output=output, \
            errors=errors: method(self, input, output, errors)
        testFunc.__doc__ = 'Parser %s: %s' % (testName, input)
        instanceMethod = new.instancemethod(testFunc, None, TestCase)
        setattr(TestCase, testName, instanceMethod)
    return unittest.TestLoader().loadTestsFromTestCase(TestCase)

def main():
    buildTestSuite()
    unittest.main()

if __name__ == "__main__":
    #Allow us to import the parent module
    os.chdir(os.path.split(os.path.abspath(__file__))[0])
    sys.path.insert(0, os.path.abspath(os.path.join(os.pardir, "src")))

    main()

File: ruby/tests/test_parser.rb
require 'test/unit'
require 'html5lib/treebuilders'
require 'html5lib/html5parser'
require 'stringio'

# add back in simpletree when errors => 0
treeTypes = {
  # "simpletree" => HTML5lib::TreeBuilders.getTreeBuilder("simpletree"),
  "rexml" => HTML5lib::TreeBuilders.getTreeBuilder("rexml")
}

puts 'Testing trees '+ treeTypes.keys.join(' ')

#Run the parse error checks
checkParseErrors = false

def parseTestcase(testString)
    testString = testString.split("\n")
    innerHTML = false
    input = []
    output = []
    errors = []
    currentList = input
    for line in testString:
        def line.startswith string; self[0...string.length] == string; end
        if line and not (line.startswith("#errors") or
          line.startswith("#document") or line.startswith("#data") or
          line.startswith("#document-fragment")):
            if currentList == output:
                if line.startswith("|"):
                    currentList.push(line[2..-1])
                else
                    currentList.push(line)
                end
            else
                currentList.push(line)
            end
        elsif line == "#errors"
            currentList = errors
        elsif line == "#document" or line.startswith("#document-fragment")
            if line.startswith("#document-fragment")
                innerHTML = line[19..-1]
                raise AssertionError unless innerHTML
            end
            currentList = output
        end
    end
    return innerHTML, input.join("\n"), output.join("\n"), errors
end

# convert the output of str(document) to the format used in the testcases
def convertTreeDump(treedump)
    treedump.split("\n")[1..-1].map {|line|
       (line.length>2 and line[0] == ?|) ? line[3..-1] : line
    }.join("\n")
end

class HTML5ParserTestCase < Test::Unit::TestCase; end

tests = 0
base = File.dirname(File.dirname(File.dirname(File.expand_path(__FILE__))))
testpath = File.join(File.join(base, 'tests'),'tree-construction')

for name, cls in treeTypes
    for filename in Dir[File.join(testpath,'*.dat')]
        f = File.open(filename)
        f.read().split("#data\n").each {|test|
            next if test == ""
            tests += 1
           
            innerHTML, input, expected, errors = parseTestcase(test)
            HTML5ParserTestCase.send :define_method, ('test_%d' % tests) do
                p = HTML5lib::HTMLParser.new(:tree => cls)
                if innerHTML
                    p.parseFragment(StringIO.new(input), innerHTML)
                else
                    p.parse(StringIO.new(input))
                end
                output = convertTreeDump(p.tree.testSerializer(p.tree.document))
                errorMsg = ["\nInput:", input,
                            "\nExpected:", expected,
                            "\nRecieved:", output].join("\n")
                assert_equal expected, output, errorMsg
                errStr = p.errors.map {|linecol,message|
                    line,col=linecol; "Line: %i Col: %i %s"%[line, col, message]}
                if checkParseErrors
                    errorMsg2 = ["\n\nInput errors:\n" + errors.join("\n"),
                        "Actual errors:\n" + errStr.join("\n")].join("\n")
                    assert_equal p.errors.length, errors.length, errorMsg2
                end
            end
        }
    end
end
